### ISSUE TITLE
Displaying arbitrary mesh

### ISSUE THREAD

**Issue by Nathipha** [Samsung/GearVRf#589]
I want to display a couple of 3D polylines (with white lines) but my tests haven't been successful yet (it doesn't display anything) and I'm now wondering what is missing. My code is:

```
float[] vertices = {1,1,1,2,2,2,3,3,3,4,4,4,5,5,5,6,6,6,7,7,7,8,8,8,9,9,9};
GVRMesh mesh = new GVRMesh(gvrContext);
GVRRenderData rd = new GVRRenderData(gvrContext);
GVRMaterial material = new GVRMaterial(gvrContext);

mesh.setVertices(vertices);
material.setColor(Color.WHITE);
rd.setMesh(mesh);
rd.setDrawMode(GLES20.GL_LINE_STRIP);
rd.setMaterial(material);

GVRSceneObject obj = new GVRSceneObject(gvrContext);
obj.getTransform().setPosition(0.0f, 0.0f, -2.0f);
obj.attachRenderData(rd);
scene.addSceneObject(obj);
```

Usually I'll get the numbers for x, y and z from a different class but I want to make it work like this before I'm using actual values.
Displaying text is quite similar to the last paragraph (set position, then add to scene) and it does work, so there's a high chance that something else is missing but what?

Is there a limit to how small/big the coordinates (everything between 0.1 and 1000 could be possible with the real numbers) should be to display everything properly?
Is it possible to change the thickness of the lines?

**Comment by thomasflynn** [comment]
@NolaDonato could you check this out?  at first glance, this looks like it should work.  thanks.

**Comment by NolaDonato** [comment]
Nathipha,

I am going to make an application from your sample code and see if I can display a polyline. I will let you know what I find.

-Nola

**Comment by NolaDonato** [comment]
I tried putting a non-textured cube in place of the polyline and it also does not display. A textured cube does display. I think something has regressed with respect to non-textured meshes. This is a problem with GearVRf and not with the sample program. I will fix it.

-Nola

**Comment by NolaDonato** [comment]
I submitted pull request #593 to fix the problems in GearVRF preventing polylines from working properly.


1.       GearVRF had a regression displaying non-textured materials (as in the textured shader will not display them at all). Currently we are checking if the main texture is ready in the textured shader and ignoring the “texture ready” problem everywhere else. My asset loader rework will take care of this by ensuring that assets are not added to the scene until all the appropriate textures have been loaded.

2.       GearVRF required indices on all meshes. I changed the renderer to just use the vertex array directly if there are no indices. This change applies to ALL shaders because I moved the mesh setup code out of the individual shaders into the renderer.

3.       There was no way to set the line width at all. The line width value is now set by specifying a value for the “line_width” uniform of your GVRMaterial. The OpenGL standard says “"Only width 1 is guaranteed to be supported; others depend on the implementation." I suspect that setting this value probably won’t give you wide lines on all phones. It did work on the Note 4.



The only issue with the example is that the default textured shadier won’t display line strips properly. The new “all-purpose” Phong shader will, however. You can use this shader by setting the shader template on the GVRRenderData:


            rd.setShaderTemplate(GVRPhongShader.class);


This tells GearVRF to generate a custom fragment and vertex shader based on your mesh and material attributes and which lights are in the scene.



Here is the revised code for your example. You need the call to setShaderTemplate but you don’t need to explicitly set the line width in the material. It will default to 1 – very thin lines. I also changed the position of your polyline because it barely is within the field of view of the camera at z = -2.


      float[] vertices = {1,1,1, 2,2,2, 3,3,3, 4,4,4, 5,5,5, 6,6,6, 7,7,7, 8,8,8, 9,9,9 };
      GVRMesh mesh = new GVRMesh(gvrContext);
      GVRRenderData rd = new GVRRenderData(gvrContext);
      GVRMaterial material = new GVRMaterial(gvrContext);

      mesh.setVertices(vertices);
      material.setColor(Color.WHITE);
      rd.setMesh(mesh);
      rd.setDrawMode(GLES20.GL_LINE_STRIP);
      material.setFloat("line_width", 8.0f);
      rd.setMaterial(material);
     rd.setShaderTemplate(GVRPhongShader.class);
      GVRSceneObject obj = new GVRSceneObject(gvrContext);
      obj.getTransform().setPosition(-2.0f, -2.0f, -8.0f);
      obj.attachRenderData(rd);
      scene.addSceneObject(obj);



The fix to GearVRF will probably go in this week. It is pull request #593 on GitHub and it is currently “Open”.  When #593 disappears from the “Open” list, you can pull over GearVRF again and try your code. Hopefully I have fixed the issue for you too.



-Nola

**Comment by Nathipha** [comment]
Thanks for adding the line width!
Also thank you for the new code, I'm going to keep an eye on the status! Am I correct in my assumption that even adding the Phong shader won't work until the fix gets merged into the master? I just tested your suggestion (only added the shader) with my current version and It didn't work and:
```
float[] vertices = {1,1,1,5,1,1,5,5,1,1,5,1};
....
rd.setDrawMode(GLES20.GL_TRIANGLE_FAN);
```
also didn't display anything.
I'm not sure what it's supposed to mean but this is what the console showed after adding the object to the scene:
SHADER: end added custom shader 1000

Do I have to set new lights myself? I don't need any shadows or actual spotlights, normal ambient lighting is fine.

Something else I noticed because I'm going to need it in the future:
Is "GL_POLYGON" ("setDrawMode(GLES20.GL_POLYGON)") not supported at the moment (I know that it's pretty slow)?

**Comment by NolaDonato** [comment]
Your assumption is correct. My fix is required for your polyline code to work.

The message just indicates that a shader had to be compiled because it wasn’t already available. We are working on supporting pre-compiled shaders so you don’t have to incur the compilation overhead for the commonly used ones.

GearVRf supports indexed triangle meshes. If your polygons are convex, they can be easily converted to triangles. Does your data include concave polygons? Is that why you need GL_POLYGON?

The way the Phong shader template works is that it adjusts to your lighting conditions by generating a custom shader. If you don’t have any lights or you don’t enable lighting, it will just use the diffuse material color.  Otherwise it will use the all the lights in the scene to illuminate your geometry.

**Comment by thomasflynn** [comment]
fyi, OpenGL ES does not support GL_POLYGON.  You will need to tessellate your data to triangles to display them.

**Comment by liaxim** [comment]
@Nathipha We have made all the necessary changes. Please get the latest framework and use the gvr-polyline sample as a reference.

**Comment by liaxim** [comment]
@Nathipha Assuming it is all good now. Planning to close the issue tomorrow.

**Comment by Nathipha** [comment]
Thank you, your changes worked, @NolaDonato!

Unfortunately there's a new problem: I have to display a couple of meshes at the same time, so I save all the GVRSceneObjects (15-20 atm) in an array and afterwards add them to the scene once all the pre-drawing work is done. The problem is that it automatically connects all of the polylines:
http://i.imgur.com/zgqfLt2.png

I checked with just one:
http://i.imgur.com/ZqgHy2k.png

... and let it get drawn without the last couple of points (just to be sure that the coordinates aren't like this) but still the same thing:
http://i.imgur.com/6YheFRn.png

If there are more than two polylines, they connect because they all get the same extra vertex:
http://i.imgur.com/xTJQtFq.png

I guess it's 0,0,0? I do set the position to (-10.0f, -1.0f, -10.0f), so it can display everything properly. Is it possible to keep it from adding/connecting to the extra point?


Ad GL_POLYGON:
My polygons can be convex or concave and there's no way of knowing which one they'll be, that's why I'm looking for something that can draw both, even if it's a bit slower.

**Comment by NolaDonato** [comment]
I am looking at your problem. In the meantime, I am sending you the sample I wrote to test the polyline code. When I run this on my phone, it spells out the word L I N E. It would help if you could send me a small sample like this that shows your problem so I can see what your code looks like.

-Nola

public class SampleScript extends GVRMain {

private GVRContext mGVRContext = null;

@Override

public void onInit(GVRContext gvrContext) {

GVRScene scene = gvrContext.getNextMainScene();

float Z = -4;

float[] L = { -2.5f, 1, Z, -2.5f, -1, Z, -1.5f, -1, Z };

float[] I = { -1, 1, Z, -1, -1, Z };

float[] N = { 0, -1, Z, 0, 1, Z, 1, -1, Z, 1, 1, Z };

float[] E = { 2.5f, 1, Z, 1.5f, 1, Z, 1.5f, -1, Z,

2.5f, -1, Z, 2.5f, 0, Z, 1.5f, 0, Z,

1.5f, 1, Z, 1.5f, -1, Z};

GVRMaterial material = new GVRMaterial(gvrContext);

GVRMesh mesh = new GVRMesh(gvrContext);

GVRRenderData rd = new GVRRenderData(gvrContext);



material.setColor(Color.WHITE);

material.setLineWidth(4.0f);



GVRSceneObject Lobj = new GVRSceneObject(gvrContext);

mesh.setVertices(L);

rd.setMesh(mesh);

rd.setDrawMode(GLES20.GL_LINE_STRIP);

rd.setMaterial(material);

rd.setShaderTemplate(GVRPhongShader.class);

Lobj.attachRenderData(rd);

scene.addSceneObject(Lobj);



GVRSceneObject Iobj = new GVRSceneObject(gvrContext);

mesh = new GVRMesh(gvrContext);

rd = new GVRRenderData(gvrContext);

mesh.setVertices(I);

rd.setMesh(mesh);

rd.setDrawMode(GLES20.GL_LINES);

rd.setMaterial(material);

rd.setShaderTemplate(GVRPhongShader.class);

Iobj.attachRenderData(rd);

scene.addSceneObject(Iobj);



GVRSceneObject Nobj = new GVRSceneObject(gvrContext);

mesh = new GVRMesh(gvrContext);

mesh.setVertices(N);

rd = new GVRRenderData(gvrContext);

rd.setMesh(mesh);

rd.setDrawMode(GLES20.GL_LINE_STRIP);

rd.setMaterial(material);

rd.setShaderTemplate(GVRPhongShader.class);

Nobj.attachRenderData(rd);

scene.addSceneObject(Nobj);



GVRSceneObject Eobj = new GVRSceneObject(gvrContext);

mesh = new GVRMesh(gvrContext);

mesh.setVertices(E);

rd = new GVRRenderData(gvrContext);

rd.setMesh(mesh);

rd.setDrawMode(GLES20.GL_LINES);

rd.setMaterial(material);

rd.setShaderTemplate(GVRPhongShader.class);

Eobj.attachRenderData(rd);

scene.addSceneObject(Eobj);

}



@Override

public void onStep() {

}



}

**Comment by Nathipha** [comment]
I just tested it with:

```
@Override
public void onInit(GVRContext gvrContext) {
    gvrc = gvrContext;
    scene = gvrContext.getNextMainScene();
    setBlock(); 
}

private void setBlock() {	
    GVRSceneObject[] meshes = new GVRSceneObject[3];

    float[] one = {7.54187f, -1.18193333f, 8.02526f,
    7.3991f, -0.3774f, 8.05622f,
    7.38362f, -0.26926667f, 8.05968f,
    7.34662f, -0.60693333f, 8.067948f,
    7.33052f, -0.46813333f, 8.07154f,
    7.31946f, -0.25526667f, 8.074016f,
    7.24949f, -0.31946667f, 8.089644f,
    7.20093f, -0.17566667f, 8.100492f,
    7.12511f, -0.03193333f, 8.11836f,
    6.30802f, -0.02846667f, 8.299964f
    };
    float[] two = {7.2713f, -0.82193333f, 7.62692f,
    7.2364f, -0.74526667f, 7.63472f,
    7.2189f, -0.64526667f, 7.6386f,
    7.2023f, -0.64526667f, 7.64232f,
    7.1849f, -0.5686f, 7.6462f,
    7.1141f, -0.38193333f, 7.662f,
    7.0967f, -0.3586f, 7.66592f,
    7.0696f, -0.3086f, 7.67196f,
    7.0618f, -0.28193333f, 7.67368f,
    7.053f, -0.2486f, 7.67564f
    };
    float[] three = {7.2143f, -0.97326667f, 7.18152f,
    7.1095f, -0.92326667f, 7.20492f,
    7.0396f, -0.88993333f, 7.22052f,
    7.0222f, -0.85326667f, 7.2244f,
    6.9707f, -0.71993333f, 7.23592f,
    6.9532f, -0.65326667f, 7.2398f,
    6.9008f, -0.47993333f, 7.25152f,
    6.8301f, -0.31993333f, 7.26728f,
    6.8135f, -0.2666f, 7.271f,
    6.7917f, -0.18993333f, 7.27588f
    };
    meshes[0] = makeMesh(one);
    meshes[1] = makeMesh(two);
    meshes[2] = makeMesh(three);
				
				
    for(int k=0;k<meshes.length;k++) {
        if(meshes[k]==null) {
            break;
        } else {
            scene.addSceneObject(meshes[k]);
        }
    }
}

private GVRSceneObject makeMesh(float[] vertices) {
    GVRSceneObject obj = new GVRSceneObject(gvrc);		
    GVRMesh mesh = new GVRMesh(gvrc);
    GVRRenderData rd = new GVRRenderData(gvrc);
    GVRMaterial material = new GVRMaterial(gvrc);
		
    mesh.setVertices(vertices);
    material.setColor(Color.WHITE);
    material.setFloat("line_width", 1.0f);
		
    rd.setMesh(mesh);
    rd.setDrawMode(GLES20.GL_LINE_STRIP);
    rd.setMaterial(material);
    rd.setShaderTemplate(GVRPhongShader.class);
    obj.getTransform().setPosition(-10.0f, -1.0f, -10.0f);
    obj.attachRenderData(rd);
		
    return obj;
}
```
(Sorry, I don't know how to hide code like you did.)
This works without showing any extra vertices, so does your "L I N E" sample.
There's a lot more going on in my onInit (get mins/maxs of all the coordinates) and setBlock because I usually get the coordinates from a txt file (then save them in arrays and extra instances of my blocks, lines and points classes).
This is how the important part of "setBlock(int b)" actually looks like:
http://pastebin.com/HaN5S3u7

Apparently "setPosition" isn't actually causing this (it still does the same thing without it), so now I have no idea what does.

Just a small, unrelated, thing:
The "setScript" method is marked as deprecated now and out of the couple of samples I tested, "SimpleSample" still uses "set Script" instead of the new "setMain", so does the new "polyline" sample. ;)
360photo, gamepad & modelviewer are fine.

**Comment by Nathipha** [comment]
One more unrelated thing:
I always had the phone plugged into my PC while working on the app to debug stuff but I also want to test the app with the Gear VR now. Is it actually possible to do this if the app isn't in the Oculus store yet?
Whenever I start it, then put the phone into the Gear VR, it automatically starts the Oculus store and disables the touchscreen.

**Comment by thomasflynn** [comment]
Two things:
- make sure you have the vr_only flag on your AndroidManifest.xml
- enable oculus developer mode.  Settings->application manager->GearVR Service->Manage Storage, tap version 7-8 times.  Then you'll get a switch to enable developer mode.

**Comment by liaxim** [comment]
Regarding setScript - it is deprecated but we have some work to do before all the samples can switch to GVRMain.

**Comment by Nathipha** [comment]
@thomasflynn 
I already had the developer mode active (otherwise it doesn't even work properly) but the flag was missing. Thank you very much, now it's working! :)

@liaxim 
Aha, okay.


Unfortunately a new problem has come up:
I can't get it to display any colors for my polylines/meshes/filled objects.

Here's some code that does display the rectangle but only with white lines/with a white filling (no matter if I use "Color.RED" or "255,0,0"):
```
float[] v = {1,1,1, 5,1,1, 5,5,1, 1,5,1};
GVRSceneObject obj = new GVRSceneObject(gvrc);		
GVRMesh mesh = new GVRMesh(gvrc);
GVRRenderData rd = new GVRRenderData(gvrc);
GVRMaterial material = new GVRMaterial(gvrc);

mesh.setVertices(v);
material.setColor(Color.RED);
//material.setColor(255, 0, 0);
material.setFloat("line_width", 4.0f);
		
rd.setMesh(mesh);
rd.setDrawMode(GLES20.GL_LINE_LOOP);
//rd.setDrawMode(GLES20.GL_TRIANGLE_FAN);
rd.setMaterial(material);
rd.setShaderTemplate(GVRPhongShader.class);

obj.getTransform().setPosition(-4.0f, -1.0f, -12.0f);
obj.attachRenderData(rd);
scene.addSceneObject(obj);
```

**Comment by NolaDonato** [comment]
Try calling material.setDiffuseColor instead of material.setColor. The phong shading model (GVRPhongShader) can operate in both lit and non-lit modes. If you have lights in your scene you will get a lit shader which uses ambient, diffuse, specular and emissive material colors. If you don't have lights, it will only use the diffuse color.
setColor is left over from the days when we had no lighting in GearVRf and the shaders just had one color component. We will eventually deprecate and remove the existing builtin shaders in favor of the more general Java-based shader generation method. I left in setColor for backwards compatibility with the existing shaders.

**Comment by Nathipha** [comment]
Thank you, setDiffuseColor did it!
About Transparency: Are there specific new methods for this too or ones I shouldn't use anymore (alpha in setDiffuseColor, setOpacity)? I know that you have to set the rendering order but does the order I add the objects to the scene also matter?

There might be a rendering problem with filled polygons:
Look at my test code from about a day ago. When I use "GL_TRIANGLE_FAN" (color doesn't matter), it draws a "standing" square just fine: http://i.imgur.com/hw6wr0H.png
But when I walk behind it (I'm using modified code from the "gamepad" sample), the square simply vanishes.

Standing square with "GL_TRIANGLE_STRIP" front: http://i.imgur.com/axbXsCh.png
Back: http://i.imgur.com/dFD6eyq.png

Lying square with "GL_TRIANGLE_FAN" from above: Nothing there.
From below: Normal square

Lying square with "GL_TRIANGLE_STRIP" from above: http://i.imgur.com/2HN23YS.png
From below: http://i.imgur.com/2qs1Jdk.png

In my app (temporarily with "GL_TRIANGLE_FAN" until I got my own algorithm for concave polygons) from above: http://i.imgur.com/4eLH8Oi.png
From below: http://i.imgur.com/kCJ4l80.png

I'm still using the Phong shader and no extra lights.

**Comment by thomasflynn** [comment]
what you're dealing with is front facing vs back facing triangles:  https://www.opengl.org/wiki/Face_Culling
it'll depend on the order in which the vertices are specified.
by default, glCullFace() is enabled.  this is generally desirable for performance reasons.  You can choose to turn it off in the GVRRenderData (with setCullFace()) of the SceneObject though.

**Comment by Nathipha** [comment]
Ah, thank you, with "setCullFace(GVRCullFaceEnum.None)" it looks a lot better now. :)

Do you possibly also know the thing about the transparency?

**Comment by thomasflynn** [comment]
sorry, i missed that.  i got focused on explaining the face culling.  @roshanch could you answer that part?  thanks.

**Comment by roshanch** [comment]
@Nathipha Could you try by setting rendering order and opacity.

**Comment by Nathipha** [comment]
@thomasflynn 
No worries. ;)

@roshanch 
I already tried it with this: http://i.imgur.com/4eLH8Oi.png
For the white polylines I set: setRenderingOrder(GVRRenderingOrder.GEOMETRY);
And for the blue polygon: setRenderingOrder(GVRRenderingOrder.TRANSPARENT) & setOpacity(0.1f)
... but no success (also tried with 1.0f for the opacity just to be sure).

It's especially a problem with GVRTextViewSceneObjects because their bounding box is covering other objects.

**Comment by NolaDonato** [comment]
If you are still using the phong shader template you need to set the alpha component of the diffuse material color



Sent from my Verizon, Samsung Galaxy smartphone


-------- Original message --------

**Comment by roshanch** [comment]
@NolaDonato  @Nathipha  GVRf uses premultiplied-alpha in blending. In phong shader, we are not taking alpha component when assigning to final color, even if we take it, it wont make an effect because of premultiplied alpha. We need to take opacity component of material and multiply it with final color just like a texture shader.

**Comment by NolaDonato** [comment]
I have been away at a conference for two days. Today I figured out where the transparency is supposed to be passed thru in the Phong shader case and it is not getting set right. I will fix it and update the polyline sample app to test transparency.

**Comment by NolaDonato** [comment]
I found the problem and tried to address it with pull request #617. I also updated the polyline sample to use transparency (pull request #202)

**Comment by Nathipha** [comment]
@NolaDonato 
Yes, I'm still using the Phong shader.
Ah, I see. Thank you for updating, I'll test it tomorrow.

**Comment by Nathipha** [comment]
Thank you very much, it's working. :)
So, are material.setOpacity() and rd.setRenderingOrder(......) even used anymore, since transparency works with the alpha component of setDiffuseColor and rd.setAlphaBlend(true) now?

Is it possible to make the black background of TextViewSceneObjects disappear with this as well somehow? According to the documentation, the alpha gets ignored for setBackgroundColor. :/

Unfortunately, if I use the alpha channel for my blue polygone, the lines from backface culling (http://i.imgur.com/kCJ4l80.png) show up again, you can see them through the semi-transparent polygone (yes, rd.setCullFace(GVRCullFaceEnum.None) is still there).

**Comment by thomasflynn** [comment]
@NolaDonato ?

**Comment by NolaDonato** [comment]
This inconsistency is because of multiple lighting support. All of our previous legacy shaders either do not use lighting or allow only one light. To support more than one light, GearVRf has to dynamically generate new shaders when you add lights to the scene. The GVRPhongShader class implements shaders which support multiple lights. This one uses the alpha component of the diffuse color.

The legacy shaders use setOpacity instead of using the alpha component of the diffuse color. They don’t support multiple input textures like the phong shader so they use the name “main_texture” for the one and only texture. The phong shader uses the names “diffuse_texture”, “specular_texture”, etc. for consistency with the corresponding material parameters “diffuse_color”, “specular_color”, …

At some point we will reimplement the legacy shaders in terms of the new infrastructure and remove the inconsistencies. For now, I did not change them when I implemented multiple lights because I didn’t want to break existing applications that used them.

setRenderingOrder should still apply – the phong shader does not behave differently in this respect.

I will look into what happens when you use the phong shader with text view scene objects. I have not tried that.

I am not sure why your lines are not being culled. I will investigate that. Do you have a small sample that shows the problem?

**Comment by Nathipha** [comment]
Ah, I see. Thanks for the explanation!

I'm not sure if it's even a problem with the Phong shader but instead simply with how the TextViewObjects are built. Try it with:
```
GVRTextViewSceneObject t = new GVRTextViewSceneObject(gvrc);
t.setText("Test");
t.setGravity(Gravity.TOP | Gravity.LEFT);
t.setTextSize(4);
scene.addSceneObject(t);
```
If you now add another object behind it to the scene (doesn't matter what), you'll notice that the white word "Test" has a small, black rectangle around it. This is what I want to get rid of or at least make transparent.
material.setDiffuseColor(1.0f, 1.0f, 1.0f, 0.5f) didn't work for me (and even if it did, I'm pretty sure it would take the whole object, not just the background), same thing with material.setOpacity(...).

About the lines:
The backface culling looks like it's working (the screenshot I added was the one from one of my other posts) but because I set the alpha to 0.5, the lines shine through it. It might me because of a concave polygon, I'm not sure but I'll do some more testing with smaller ones.

Sorry, I've got another question:
I have a jpg (I know the size and the coordinates for the corners) I want to display with my other objects. I know that you can simply make a GVRSceneObject with the length, width and GVRBitmapTexture but getting it into place through the right rotations is a little bit more complicated.
If I now make my own SceneObject (let's say, with {1,1,1, 5,1,1, 5,5,1, 1,5,1}, it will be a rectangle or square for sure!) and fill it with "GLES20.GL_TRIANGLE_FAN", is it possible to map the texture onto the filled part (didn't work for me with "material.setMainTexture(...)")?

**Comment by NolaDonato** [comment]
Did you put texture coordinates on the mesh that you are trying to texture map? Here is an example:
        GVRMesh mesh = new GVRMesh(this);
        float width = 1.0f;
        float height = 1.0f;
        float[] vertices = { width * -0.5f, height * 0.5f, 0.0f, width * -0.5f,
                height * -0.5f, 0.0f, width * 0.5f, height * 0.5f, 0.0f,
                width * 0.5f, height * -0.5f, 0.0f };
        mesh.setVertices(vertices);

        final float[] normals = { 0.0f, 0.0f, 1.0f, 0.0f, 0.0f, 1.0f, 0.0f,
                0.0f, 1.0f, 0.0f, 0.0f, 1.0f };
        mesh.setNormals(normals);

        final float[] texCoords = { 0.0f, 0.0f, 0.0f, 1.0f, 1.0f, 0.0f, 1.0f, 1.0f };
        mesh.setTexCoords(texCoords);

        char[] triangles = { 0, 1, 2, 1, 3, 2 };
        mesh.setTriangles(triangles);

        return mesh;

**Comment by Nathipha** [comment]
Thanks, I forgot about those.
setTriangles is deprecated now and it causes weird artifacts:
http://i.imgur.com/PkBuPCV.png

Unforunately it looks like it doesn't map it properly because of the "TRIANGLE_FAN" draw mode, am I missing anything?
Looks like this:
http://i.imgur.com/LHYyIWf.png
Should be this:
https://i.ytimg.com/vi/9udYi7exojk/maxresdefault.jpg

Is there a way to automatically resize the picture to fill the whole plane using the same vertices as for the same mesh but without the y (instead of 0,0, 0,1,...)?
If I try it that way, it looks like this at the moment:
http://i.imgur.com/erJ5P6s.png

And if I use the Phong shader, the plane is simply white.

It maps it properly with
    GVRSceneObject obj = new GVRSceneObject(gvrc, 4.0f, 2.0f, bmt);
instead of
    material.setMainTexture(bitmaptexture);
(no matter is standard or Phong shader) but then I can't use my own coordinates of course.

**Comment by NolaDonato** [comment]
With the phong shader you need  to do material.setTexture("diffuseTexture", bitmaptexture). It doesn't name the textures the same: it uses "diffuseTexture", "ambientTexture", ... for the names because it supports multiple textures.

**Comment by NolaDonato** [comment]
TRIANGLE_FAN assumes the first vertex is the same for all the triangles. TRIANGLES takes each set of three vertices as a different triangle. Which way are your vertices set up? Maybe you want TRIANGLES?
http://stackoverflow.com/questions/20394727/gl-triangle-strip-vs-gl-triangle-fan

**Comment by Nathipha** [comment]
Ah, the "diffuseTexture" part did the trick for the PhongShader now, thanks. It now looks like my "Looks like this" picture but without the artifact overlay.

I used TRIANGLE_FAN to fill the plane, so I can map my texture onto it. Since it's a simple square (so with four 90° angles, my array for mesh.setVertices(...) contains the 12 x, y and z coordinates of the 4 corner points) and GL_QUADS isn't supported, shouldn't that work for it? At least the filling part works but the mapping the texture correctly one not so much.
I think I know why it maps the texture the way it does: Because of how TRIANGLE_FAN splits the square into triangles. But is there a way to avoid that?

Btw, I got concave polygons to work yesterday by using an extra algorithm that gives me the coordinates as already correct triangles and then I simply fill everything with GL_TRIANGLES. Now the weird lines from my older post (11 days ago, 27th?) are gone too. :)

**Comment by thomasflynn** [comment]
Gl_QUADS is not supported by GLES.  I'll let @Noladonato handle the rest of your question.

**Comment by NolaDonato** [comment]
I will be away for a week, I think Roshan can help you while I am gone.

**Comment by NolaDonato** [comment]
Is this resolved? Do you need further help on this issue?

**Comment by liaxim** [comment]
@Nathipha If you don't have any other comments regarding this issue I will close it tomorrow. Thanks.

**Comment by Nathipha** [comment]
Sorry for my late answer.
The latest problem with mapping textures on a plane that got filled with GL_TRIANGLE_FAN incorrectly might still be there (haven't had the time to test it in 3.0.1 yet) but there's an alternative I've been using:

    GVRSceneObject obj = new GVRSceneObject(gvrc, 4.0f, 2.0f, bitmaptexture);

This way it's more complicated because you can't simply build the plane whereever you want it but you have to move and turn it instead but it does map textures correctly, so I'll just keep it.
Thanks for your help, everyone!
